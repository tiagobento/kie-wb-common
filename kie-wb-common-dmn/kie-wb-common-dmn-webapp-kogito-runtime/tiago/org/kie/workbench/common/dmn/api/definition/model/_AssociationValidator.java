package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _AssociationValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.Association> {
  static _AssociationValidator INSTANCE = GWT.create(_AssociationValidator.class);
}
