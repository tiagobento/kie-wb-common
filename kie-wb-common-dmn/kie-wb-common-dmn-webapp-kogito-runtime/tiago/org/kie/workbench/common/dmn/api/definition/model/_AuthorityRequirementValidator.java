package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _AuthorityRequirementValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.AuthorityRequirement> {
  static _AuthorityRequirementValidator INSTANCE = GWT.create(_AuthorityRequirementValidator.class);
}
