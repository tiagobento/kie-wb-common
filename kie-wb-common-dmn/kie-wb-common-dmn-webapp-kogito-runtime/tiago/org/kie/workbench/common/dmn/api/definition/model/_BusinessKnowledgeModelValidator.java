package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _BusinessKnowledgeModelValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.BusinessKnowledgeModel> {
  static _BusinessKnowledgeModelValidator INSTANCE = GWT.create(_BusinessKnowledgeModelValidator.class);
}
