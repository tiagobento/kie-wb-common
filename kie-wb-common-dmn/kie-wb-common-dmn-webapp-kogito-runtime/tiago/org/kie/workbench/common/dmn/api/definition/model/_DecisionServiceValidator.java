package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _DecisionServiceValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.DecisionService> {
  static _DecisionServiceValidator INSTANCE = GWT.create(_DecisionServiceValidator.class);
}
