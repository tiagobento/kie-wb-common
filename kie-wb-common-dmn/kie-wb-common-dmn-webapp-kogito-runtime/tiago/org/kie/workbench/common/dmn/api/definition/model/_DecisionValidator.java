package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _DecisionValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.Decision> {
  static _DecisionValidator INSTANCE = GWT.create(_DecisionValidator.class);
}
