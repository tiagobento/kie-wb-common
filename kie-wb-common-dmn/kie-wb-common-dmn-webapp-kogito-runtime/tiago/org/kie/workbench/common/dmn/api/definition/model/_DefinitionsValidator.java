package org.kie.workbench.common.dmn.api.definition.model;

import com.google.gwt.core.client.GWT;

public interface _DefinitionsValidator extends com.google.gwt.validation.client.impl.GwtSpecificValidator <org.kie.workbench.common.dmn.api.definition.model.Definitions> {
  static _DefinitionsValidator INSTANCE = GWT.create(_DefinitionsValidator.class);
}
