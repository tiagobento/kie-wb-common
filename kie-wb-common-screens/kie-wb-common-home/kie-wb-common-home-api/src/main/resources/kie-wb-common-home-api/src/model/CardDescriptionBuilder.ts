import {
  CardDescription,
  CardDescriptionElement,
  CardDescriptionLinkElement,
  CardDescriptionTextElement
} from "./CardDescription";

export class CardDescriptionBuilder {
  private static readonly linkTokenRegex = new RegExp(/{(\d+)}/g);

  private readonly textMask: string;
  private readonly links: CardDescriptionLinkElement[];

  constructor(textMask: string) {
    this.textMask = textMask;
    this.links = [];
  }

  public addLinkIf(predicate: () => boolean, text: string, targetId: string): CardDescriptionBuilder {
    if (!predicate()) {
      return this;
    }
    return this.addLink(text, targetId);
  }

  public addLink(text: string, targetId: string): CardDescriptionBuilder {
    this.links.push(new CardDescriptionLinkElement(text, targetId));
    return this;
  }

  public build(): CardDescription {
    const linkTokens = this.textMask.match(CardDescriptionBuilder.linkTokenRegex);

    if (linkTokens === null || linkTokens.length === 0) {
      return new CardDescription([new CardDescriptionTextElement(this.textMask)]);
    }

    const elements: CardDescriptionElement[] = [];
    for (let i = 0; i < linkTokens.length; i++) {
      const textBeforeLink = this.getTextUntilLinkAt(this.textMask, linkTokens, i);

      if (textBeforeLink.length > 0) {
        elements.push(new CardDescriptionTextElement(textBeforeLink));
      }

      elements.push(this.links[this.linkIdxFromToken(linkTokens[i])]);
    }

    const lastElement = new CardDescriptionTextElement(
      this.getTextUntilLinkAt(this.textMask, linkTokens, linkTokens.length)
    );

    if (lastElement.text.length > 0) {
      elements.push(lastElement);
    }

    return new CardDescription(elements);
  }

  private getTextUntilLinkAt(textMask: string, tokens: RegExpMatchArray, i: number) {
    const prevLinkIdx = i === 0 ? 0 : this.linkIdxFromToken(tokens[i - 1]);
    const prevLinkTkn = `{${prevLinkIdx}}`;

    const textTokenStart = i === 0 ? 0 : textMask.indexOf(prevLinkTkn) + prevLinkTkn.length;
    const textTokenEnd = i === tokens.length ? textMask.length : textMask.indexOf(tokens[i]);

    return textMask.substring(textTokenStart, textTokenEnd);
  }

  private linkIdxFromToken(tkn: string) {
    return parseInt(tkn.replace("{", "").replace("}", ""), 10);
  }
}
