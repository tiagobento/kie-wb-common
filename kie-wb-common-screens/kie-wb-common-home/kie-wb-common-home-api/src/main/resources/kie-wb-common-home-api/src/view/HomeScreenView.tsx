import * as React from "react";
import {
  Profile,
  ProfilePreferences,
  ProfilePreferencesPortableGeneratedImpl
} from "@kiegroup-ts-generated/kie-wb-common-profile-api";
import { PreferenceBeanServerStore } from "@kiegroup-ts-generated/uberfire-preferences-api-rpc";
import { CardView } from "./CardView";
import { HomeScreen, HomeScreenProvider } from "../model";

interface Props {
  contentProvider: HomeScreenProvider;
}

interface State {
  model?: HomeScreen;
}

export class HomeScreenView extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {};

    this.retrieveCurrentProfile(new PreferenceBeanServerStore())
      .then(profile => {
        this.setState({ model: props.contentProvider.get(profile) });
      })
      .catch(() => new Error("Could not retrieve profile preferences."));
  }

  public render() {
    if (!this.state.model) {
      return <></>;
    }

    const containerStyle = { backgroundImage: `url(${this.state.model.backgroundImageUrl})` };

    // TODO: link with HomeView.less
    return (
      <div id="home-page">
        <div className="kie-page">
          <div
            data-field="container"
            className="kie-page__content kie-content--bg-image kie-blank-slate"
            style={containerStyle}
          >
            <div className="container-fluid kie-container-fluid--blank-slate">
              <div className="blank-slate-pf row">
                <h1 data-field="welcome">{this.state.model.welcomeText}</h1>

                <p data-field="description">{this.state.model.description}</p>

                <div data-field="shortcuts" className="blank-slate-pf-secondary-action">
                  {this.state.model.cards.map((card, idx) => (
                    <CardView model={card} key={idx} />
                  ))}
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }

  private retrieveCurrentProfile(preferencesStore: PreferenceBeanServerStore): Promise<Profile> {
    const args = { emptyPortablePreference: new ProfilePreferencesPortableGeneratedImpl({}) };

    return preferencesStore
      .load2<ProfilePreferences, ProfilePreferencesPortableGeneratedImpl>(args)
      .then(pref => pref.profile!);
  }
}
