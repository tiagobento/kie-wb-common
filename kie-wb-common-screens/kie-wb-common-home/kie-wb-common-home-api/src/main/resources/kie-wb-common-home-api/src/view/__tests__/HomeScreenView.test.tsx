import { Card, CardDescriptionBuilder, HomeScreen, HomeScreenProvider } from "../../model";
import { Profile, ProfilePreferencesPortableGeneratedImpl } from "@kiegroup-ts-generated/kie-wb-common-profile-api";
import { shallow } from "enzyme";
import * as React from "react";
import { HomeScreenView } from "../HomeScreenView";
import { PreferenceBeanServerStore } from "@kiegroup-ts-generated/uberfire-preferences-api-rpc";

describe("behaviour", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("with a profile, should call provider passing the correct profile", async () => {
    const provider = new FooHomeScreenProvider();

    const getSpy = jest.spyOn(provider, "get");

    const pref = new ProfilePreferencesPortableGeneratedImpl({ profile: Profile.FULL });
    const promise = Promise.resolve(pref);

    PreferenceBeanServerStore.prototype.load2 = jest.fn(() => promise);

    const component = shallow(<HomeScreenView contentProvider={provider} />);

    await Promise.all([promise]);

    expect(PreferenceBeanServerStore.prototype.load2).toHaveBeenCalledTimes(1);
    expect(getSpy).toHaveBeenCalledWith(Profile.FULL);

    expect(component.prop("id")).toEqual("home-page");
    expect(component.children()).toHaveLength(1);

    const kiePage = component.childAt(0);
    expect(kiePage.prop("className")).toEqual("kie-page");
    expect(kiePage.children()).toHaveLength(1);

    const background = kiePage.childAt(0);
    expect(background.prop("data-field")).toEqual("container");
    expect(background.prop("className")).toEqual("kie-page__content kie-content--bg-image kie-blank-slate");
    expect(background.children()).toHaveLength(1);
    expect(background.prop("style")).toStrictEqual({
      backgroundImage: "url(foo_bg.png)"
    });

    const container = background.childAt(0);
    expect(container.prop("className")).toEqual("container-fluid kie-container-fluid--blank-slate");
    expect(container.children()).toHaveLength(1);

    const row = container.childAt(0);
    expect(row.prop("className")).toEqual("blank-slate-pf row");
    expect(row.children()).toHaveLength(3);

    const welcomeElement = row.childAt(0);
    expect(welcomeElement.prop("data-field")).toEqual("welcome");
    expect(welcomeElement.text()).toEqual("welcome!");

    const descriptionElement = row.childAt(1);
    expect(descriptionElement.prop("data-field")).toEqual("description");
    expect(descriptionElement.text()).toEqual("this is a description");

    const shortcutsElement = row.childAt(2);
    expect(shortcutsElement.prop("data-field")).toEqual("shortcuts");
    expect(shortcutsElement.prop("className")).toEqual("blank-slate-pf-secondary-action");
    expect(shortcutsElement.children()).toHaveLength(2);

    const homeScreenContent = new FooHomeScreenProvider().get(Profile.FULL);

    const firstCardElement = shortcutsElement.childAt(0);
    expect(firstCardElement.key()).toEqual("0");
    expect(firstCardElement.prop("model")).toEqual(homeScreenContent.cards[0]);

    const secondCardElement = shortcutsElement.childAt(1);
    expect(secondCardElement.key()).toEqual("1");
    expect(secondCardElement.prop("model")).toEqual(homeScreenContent.cards[1]);
  });
});

describe("snapshot", () => {
  test("with a profile, should render consistently", async () => {
    const provider = new FooHomeScreenProvider();

    const pref = new ProfilePreferencesPortableGeneratedImpl({ profile: Profile.FULL });
    const promise = Promise.resolve(pref);

    PreferenceBeanServerStore.prototype.load2 = jest.fn(() => promise);

    const component = shallow(<HomeScreenView contentProvider={provider} />);

    await Promise.all([promise]);

    expect(component).toMatchSnapshot();
  });
});

class FooHomeScreenProvider implements HomeScreenProvider {
  public get(profile: Profile): HomeScreen {
    const welcomeText = "welcome!";
    const description = "this is a description";
    const backgroundImageUrl = "foo_bg.png";

    const cardDescription = new CardDescriptionBuilder("main card description").build();
    const card1 = new Card([], "first card", cardDescription, "main-perspective");
    const card2 = new Card([], "second card", cardDescription, "main-perspective");

    return new HomeScreen(welcomeText, description, backgroundImageUrl, [card1, card2]);
  }
}
