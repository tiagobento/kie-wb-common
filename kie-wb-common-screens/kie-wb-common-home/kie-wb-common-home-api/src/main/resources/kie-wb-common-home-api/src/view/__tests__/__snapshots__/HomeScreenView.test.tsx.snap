// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot with a profile, should render consistently 1`] = `
<div
  id="home-page"
>
  <div
    className="kie-page"
  >
    <div
      className="kie-page__content kie-content--bg-image kie-blank-slate"
      data-field="container"
      style={
        Object {
          "backgroundImage": "url(foo_bg.png)",
        }
      }
    >
      <div
        className="container-fluid kie-container-fluid--blank-slate"
      >
        <div
          className="blank-slate-pf row"
        >
          <h1
            data-field="welcome"
          >
            welcome!
          </h1>
          <p
            data-field="description"
          >
            this is a description
          </p>
          <div
            className="blank-slate-pf-secondary-action"
            data-field="shortcuts"
          >
            <CardView
              key="0"
              model={
                Card {
                  "description": CardDescription {
                    "elements": Array [
                      CardDescriptionTextElement {
                        "text": "main card description",
                      },
                    ],
                  },
                  "iconCssClasses": Array [],
                  "onMayClick": undefined,
                  "perspectiveId": "main-perspective",
                  "title": "first card",
                }
              }
            />
            <CardView
              key="1"
              model={
                Card {
                  "description": CardDescription {
                    "elements": Array [
                      CardDescriptionTextElement {
                        "text": "main card description",
                      },
                    ],
                  },
                  "iconCssClasses": Array [],
                  "onMayClick": undefined,
                  "perspectiveId": "main-perspective",
                  "title": "second card",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
